Forgot password 3

divy-arpit opened the pull request:
## Fixes #6 

### Description

Added forgot password cubit along with fake repo

**Type of Change:** (Delete irrelevant options)
- Code


**Code/Quality Assurance Only** (Delete irrelevant options)

- New feature (non-breaking change which adds functionality pre-approved by mentors)


### How Has This Been Tested?

Tested on local machine.


### Checklist:

- [X] My PR follows the style guidelines of this project
- [X] I have performed a self-review of my own code or materials
- [ ] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] Any dependent changes have been merged
- [ ] I have added one/more required dependencies
- [X] I have formatted the code using dart format (dartfmt)


**Code/Quality Assurance**
- [X] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

**Mention the dependencies/packages/plugins that have been added (if any)**

divy-arpit committed:
added forgot password cubit

divy-arpit committed:
removed changes from pubspec.lock

divy-arpit committed:
update

divy-arpit committed:
update

divy-arpit committed:
update

divy-arpit committed:
removed un used variables

werainkhatri commented on lib/screens/forgot_password/cubit/forgot_password_cubit.dart:
```diff
@@ -0,0 +1,32 @@
+
+  ForgotPasswordCubit(this._forgotPasswordRepository) : super(ForgotPasswordInitial());
+
+  void generateOTPInCubit(String email) async {
```
1. give better function names. This is supposed to initiate the forgot password process, so calling it `sendOTP` or `forgotPassword` or similar is preferable. You are not generating the otp in cubit, so it is a misleading name.
2. Should return `Future<void>`

werainkhatri replied:
2 is undone

werainkhatri commented on lib/screens/forgot_password/forgot_password_repo.dart:
```diff
@@ -0,0 +1,33 @@
+import 'dart:math';
```
🔴 
file name should be complete - `forgot_password_repository.dart`. Don't use abbreviations.

werainkhatri commented on lib/screens/forgot_password/forgot_password_repo.dart:
```diff
@@ -0,0 +1,33 @@
+abstract class ForgotPasswordRepository {
+  Future<int> generateOTP(String email);
+
+  Future<bool> verifyOTP(String x);
```
give better variable names.
```suggestion
  Future<bool> verifyOTP(String otp);
```

werainkhatri replied:
add doc comments for both the functions.

werainkhatri commented on lib/screens/forgot_password/forgot_password_repo.dart:
```diff
@@ -0,0 +1,33 @@
+      throw NetworkException();
+    } else {
+      otp = 1234;
+      return 202; // code for successful generation of otp
```
any specific reason for returning int instead of just void?

werainkhatri commented on lib/screens/forgot_password/forgot_password_repo.dart:
```diff
@@ -0,0 +1,33 @@
+}
+
+class FakeForgotPasswordRepository implements ForgotPasswordRepository {
+  int otp;
```
reason for this variable?

werainkhatri commented on lib/screens/forgot_password/forgot_password_repo.dart:
```diff
@@ -0,0 +1,33 @@
+    }
+  }
+
+  Future<bool> verifyOTP(String x) async {
```
this should be overriden

werainkhatri commented on lib/screens/forgot_password/cubit/forgot_password_cubit.dart:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  Future<void> verifyOTP(String a) async {
+    bool b = await _forgotPasswordRepository.verifyOTP(a);
```
try catch for this?

werainkhatri commented on lib/screens/forgot_password/cubit/forgot_password_state.dart:
```diff
@@ -0,0 +1,20 @@
+part of 'forgot_password_cubit.dart';
+
+@immutable
+abstract class ForgotPasswordState {}
```
should extend Equatable

werainkhatri commented on lib/screens/forgot_password/cubit/forgot_password_state.dart:
```diff
@@ -0,0 +1,20 @@
+@immutable
+abstract class ForgotPasswordState {}
+
+class ForgotPasswordInitial extends ForgotPasswordState {}
```
Since this is a multi step screen, the states should also be for each step. Eg:
- `ForgotEmailInitial`, `ForgotEmailError`, `ForgotEmailSuccess`, `ForgotEmailFailure`
- `ForgotOTPInitial`, `ForgotOTPError`, `ForgotOTPSuccess`, `ForgotOTPFailure`
- `ForgotChangeInitial`, ...
- `ForgotLoading` cause loading is same for all steps.

removed **Password** from the names coz they are too long.

werainkhatri commented on lib/screens/forgot_password/forgot_password.dart:
```diff
@@ -0,0 +1,108 @@
+    );
+  }
+
+  Widget _initialForgotPasswrod(BuildContext context) {
```
🔴 
Spelling Mistake

werainkhatri reviewed (changes requested):
TODO
- Add enter new password step (Should have 1 fields, new password and confirm password)
- Create a steps folder that will contain stateless (/stateful if needed) widgets for each step.

divy-arpit committed:
added change password state

this commit adds  a change password state and also removes unnecessary variables.

werainkhatri reviewed (commented):
Some of the above changes aren't done yet. Please complete them

divy-arpit committed:
update on states

werainkhatri commented on lib/screens/forgot_password/forgot_password_repository.dart:
```diff
@@ -0,0 +1,43 @@
+  Future<void> changePassword(String email, String otp, String password) async {
+    print(email);
+    print(otp);
+    print(password);
```
1. use Log.d() for debugging.
2. remove them when committing.

werainkhatri commented on lib/screens/forgot_password/forgot_password_repository.dart:
```diff
@@ -0,0 +1,43 @@
+  Future<void> changePassword(String email, String otp, String password);
+}
+
+// To emulate a fake process for forgot password
```
🔴 
doc comments start with `///` and not `//`.

werainkhatri commented on lib/screens/forgot_password/forgot_password_repository.dart:
```diff
@@ -0,0 +1,43 @@
+// To emulate a fake process for forgot password
+class FakeForgotPasswordRepository implements ForgotPasswordRepository {
+  @override
+  // this is to simulate a delay for getting otp
```
1. doc comments for functions that are overriden should be done in their abstract definition (in the `ForgotPasswordRepository`), not its children.
2. also should be `///`, not `//`

werainkhatri replied:
This is not done yet.

werainkhatri commented on lib/screens/forgot_password/forgot_password_repository.dart:
```diff
@@ -0,0 +1,43 @@
+}
+
+// To emulate a fake process for forgot password
+class FakeForgotPasswordRepository implements ForgotPasswordRepository {
```
extend it instead of implementing.
```suggestion
class FakeForgotPasswordRepository extends ForgotPasswordRepository {
```

werainkhatri commented on lib/screens/forgot_password/forgot_password.dart:
```diff
@@ -1,55 +1,54 @@
       body: BlocConsumer<ForgotPasswordCubit, ForgotPasswordState>(
         listener: (context, state) {
-          if (state is ForgotPasswordError) {
+          if (state is ForgotNetworkError) {
```
Since `ForgotNetworkError` and `ForgotWrongOTP` are similar (snackbar with Text), we should have one state - `ForgotError` which has an error message.

werainkhatri commented on lib/screens/forgot_password/forgot_password.dart:
```diff
@@ -1,55 +1,54 @@
+          } else if (state is ForgotEnterOTP) {
             return _enterOTP(context);
-          } else if (state is ForgotPasswordWrongOTP) {
+          } else if (state is ForgotWrongOTP) {
```
wrongotp is not supposed to change the screen. also, in the above review, I have told you to remove it. So remove it from here too.

werainkhatri reviewed (changes requested):
Some changes.

divy-arpit committed:
made dedicated states for each step of forgot password

werainkhatri commented on lib/screens/forgot_password/cubit/forgot_password_state.dart:
```diff
@@ -8,27 +8,39 @@ abstract class ForgotPasswordState extends Equatable {
+class ForgotEmailInitial extends ForgotPasswordState {}
+
+/// Success state for email
+class ForgotEmailSuccess extends ForgotPasswordState {}
```
I realise this and OTPSucess states are not required. remove them.

werainkhatri commented on lib/screens/forgot_password/cubit/forgot_password_state.dart:
```diff
@@ -8,27 +8,39 @@ abstract class ForgotPasswordState extends Equatable {
+
+/// network error for forgot password
 class ForgotNetworkError extends ForgotPasswordState {
   final String error;
```
```suggestion
  final String message;
```

werainkhatri commented on lib/screens/forgot_password/forgot_password.dart:
```diff
@@ -21,24 +21,36 @@ class ForgotPasswordScreen extends StatelessWidget {
-          } else if (state is ForgotPasswordSuccess) {
+          } else if (state is ForgotResetSuccess) {
             return _passwordResetSuccess();
           } else {
```
There is a problem here. consider this scenario: 
User enters their email, moves to the otp screen, the internet goes off, and they try to check their otp. A network error occurs, the code runs this else function and the user goes to the enter email screen. We'll have a call over how you can solve this issue.

divy-arpit replied:
Okay. Let's discuss this over a call.

divy-arpit committed:
fixed the docs comment error and other minor errors

werainkhatri commented:
Update on this PR?

divy-arpit committed:
combined all network errors into one state

all network errors for various states have been combined into a single error state.

divy-arpit committed:
Added all the response codes according to the backend docs

this commit adds all the responses which could be fetched from the backend to the fake and handles them accordingly.

divy-arpit commented:
@werainkhatri done

divy-arpit committed:
converted error code classes to throwable errors

all the error codes have been removed and now all the errors would be thrown.

divy-arpit committed:
removed some dead code

werainkhatri commented on lib/screens/forgot_password/forgot_password_repository.dart:
```diff
@@ -3,42 +3,38 @@ import 'dart:math';
+  Future<void> sendOTP(String email) async {
     await Future.delayed(Duration(seconds: 2));
     if (Random().nextBool()) {
       return "200";
```
no need to return anything when return type is void.
```suggestion
      return;
```

werainkhatri commented on lib/screens/forgot_password/forgot_password_repository.dart:
```diff
@@ -3,42 +3,38 @@ import 'dart:math';
 // this to simulate the process of otp verification
   @override
-  Future<String> verifyOTP(String otp) async {
+  Future<void> verifyOTP(String otp) async {
```
1. as per the API requirements, this function needs String otp and String email.
2. to prevent confusions, name it `checkOTP` as `check_otp` is the api endpoint.

divy-arpit committed:
removed a small bug

werainkhatri commented on lib/screens/forgot_password/cubit/forgot_password_cubit.dart:
```diff
@@ -15,43 +15,37 @@ class ForgotPasswordCubit extends Cubit<ForgotPasswordState> {
       emit(ForgotNetworkError(S.networkException, state));
+    } on ResponseException catch (e) {
+      emit(ForgotOTPFailure(e.message, state));
     }
```
Here, and in other functions below, add a `on catch(e)` with `Log.s`. Refer other cubits how it was done there.

werainkhatri commented on lib/screens/forgot_password/cubit/forgot_password_cubit.dart:
```diff
@@ -12,37 +12,40 @@ class ForgotPasswordCubit extends Cubit<ForgotPasswordState> {
   }
 
-  Future<void> verifyOTP(String a) async {
+  Future<void> verifyOTP(String a, ForgotPasswordState state) async {
```
1. give better variable names.
2. as per the API requirements, this function needs String otp and String email.
3. to prevent confusions, name it `checkOTP` as `check_otp` is the api endpoint.

werainkhatri commented on lib/screens/forgot_password/cubit/forgot_password_state.dart:
```diff
@@ -12,20 +12,22 @@ class ForgotLoading extends ForgotPasswordState {}
 class ForgotOTPInitial extends ForgotPasswordState {}
 
-class ForgotOTPFailure extends ForgotPasswordState {}
+class ForgotOTPFailure extends ForgotPasswordState {
```
Since Error is used everywhere as a convention now.
```suggestion
class ForgotPasswordError extends ForgotPasswordState {
```

werainkhatri commented on lib/screens/forgot_password/cubit/forgot_password_cubit.dart:
```diff
@@ -12,37 +12,40 @@ class ForgotPasswordCubit extends Cubit<ForgotPasswordState> {
   }
 
-  Future<void> changePassword(String email, String otp, String password) async {
+  Future<void> changePassword(
+      String email, String otp, String password, ForgotPasswordState state) async {
```
Since logic is supposed to be implemented in cubit, accept both passwords and check here if they are equal first before calling repository `changePassword` function.
If they are equal, pass the `newPassword` variable to the repository
If not, emit a `ForgotPasswordError` state and return.
```suggestion
  Future<void> changePassword(
      String email, 
      String otp, 
      String newPassword,
      String confirmPassword, 
      ForgotPasswordState state) async {
```

werainkhatri replied:
This is remaining too.

werainkhatri reviewed (changes requested):
Test it once from your side by running it to see if it works at all the steps.

divy-arpit committed:
made the request changes

divy-arpit committed:
changed name a per convention

werainkhatri commented on lib/screens/forgot_password/cubit/forgot_password_cubit.dart:
```diff
@@ -0,0 +1,62 @@
+      emit(ForgotLoading());
+      await _forgotPasswordRepository.sendOTP(email);
+      emit(ForgotOTPInitial());
+    } on NetworkException {
+      emit(ForgotNetworkError(S.networkException, state));
+    } on ResponseException catch (e) {
+      emit(ForgotPasswordError(e.message, state));
```
Technically, there is no need for ForgotNetworkError
```suggestion
    } on NetworkException {
      emit(ForgotPasswordError(S.networkException, state));
    } on ResponseException catch (e) {
      emit(ForgotPasswordError(e.message, state));
```
1. treat ForgotNetworkError same as ForgotPasswordError
2. remove ForgotNetworkError state
3. do the above changes in other functions too.

werainkhatri reviewed (changes requested):
Good work. final changes.

divy-arpit committed:
replaced forgot network error state with forgot password error

werainkhatri reviewed (approved):
Great work!

werainkhatri committed:
Merge branch 'main' into forgot-password-3
